mongo/integration/client_test.go

```diff
@@ -1006,3 +1008,137 @@ func TestClientStress(t *testing.T) {
 		}
 	})
 }
+
+func TestCSOT(t *testing.T) {
+	mt := mtest.New(t, mtest.NewOptions().CreateClient(false))
+
+	csotOpts := mtest.NewOptions().ClientOptions(options.Client().SetTimeout(10 * time.Second))
+	mt.RunOpts("includes maxTimeMS if CSOT timeout is set", csotOpts, func(mt *mtest.T) {
+		mt.Run("with context.Background", func(mt *mtest.T) {
+			_, err := mt.Coll.InsertOne(context.Background(), bson.D{})
+			require.NoError(mt, err, "InsertOne error")
+
+			maxTimeVal := mt.GetStartedEvent().Command.Lookup("maxTimeMS")
+
+			require.True(mt, len(maxTimeVal.Value) > 0, "expected maxTimeMS BSON value to be non-empty")
+			require.Equal(mt, maxTimeVal.Type, bsontype.Int64, "expected maxTimeMS value to be type Int64")
+
+			maxTimeMS := maxTimeVal.Int64()
+			assert.True(mt, maxTimeMS > 0, "expected maxTimeMS value to be greater than 0")
+		})
+		mt.Run("with context.WithTimeout", func(mt *mtest.T) {
+			ctx, cancel := context.WithTimeout(context.Background(), 10*time.Minute)
+			defer cancel()
+
+			_, err := mt.Coll.InsertOne(ctx, bson.D{})
+			require.NoError(mt, err, "InsertOne error")
+
+			maxTimeVal := mt.GetStartedEvent().Command.Lookup("maxTimeMS")
+			require.True(mt, len(maxTimeVal.Value) > 0, "expected maxTimeMS BSON value to be non-empty")
+			require.Equal(mt, maxTimeVal.Type, bsontype.Int64, "expected maxTimeMS value to be type Int64")
+
+			maxTimeMS := maxTimeVal.Int64()
+			assert.True(mt,
+				maxTimeMS > 60_000,
+				"expected maxTimeMS value to be greater than 60000, but got %v",
+				maxTimeMS)
+		})
+	})
+
+	mt.RunOpts("timeout errors wrap context.DeadlineExceeded", csotOpts, func(mt *mtest.T) {
+		// Test that a client-side timeout is a context.DeadlineExceeded
+		mt.Run("MaxTimeMSExceeded", func(mt *mtest.T) {
```

Comment:
This test is failing and seems related to [this](https://github.com/mongodb/mongo-go-driver/pull/1588#discussion_r1546649121) thread.

Reply:
Fixed.